### ISSUE TITLE
Multiple endpoints with ALAG in FOCEI estimation problem

### ISSUE THREAD

**Issue by jpryby** [nlmixrdevelopment/nlmixr#429]
Running: nlmixr_1.1.1-9, RxODE_0.9.2-2 and n1qn1_6.0.1-9

I have a model with multiple endpoints in which both ALAG and F need to be estimated, but there is no sensitivity on either of those variables. The model runs fine if I do one of the following:

- ALAG is removed,
- Only one endpoint is predicted, or
- FO is used

Otherwise, the following error message occurs:
```
Error in .fitFun(.ret) : 
  Evaluation error: Infinite/NaN while evaluating initial objective function.
Restart 1
Error in .fitFun(.ret) : 
  Not compatible with requested type: [type=list; target=double].
Restart 2
Error in .fitFun(.ret) : 
  Not compatible with requested type: [type=list; target=double].
Restart 3
Error in .fitFun(.ret) : 
  Not compatible with requested type: [type=list; target=double].
Error in (function (data, inits, PKpars, model = NULL, pred = NULL, err = NULL,  : 
  Could not fit data.
```
F doesn't seem to be causing any problems, but I figured it should be mentioned since it's in the same "family" of parameters.

I can't share the model, but I was able to reproduce the issue with the warfarin model you've used for previous demos:
``` r
library(nlmixr)
pk.turnover.emax.lag <- function() {
  ini({
    talag <- c(0,0.1)
    tka <- log(1)
    tcl <- log(0.1)
    tv <- log(10)
    ##
    eta.ka ~ 1
    eta.cl ~ 2
    eta.v ~ 1
    prop.err <- 0.1
    pkadd.err <- 0.1
    ##
    poplogit <- 2
    #temax <- 7.5
    tec50 <- log(0.5)
    tkout <- log(0.05)
    te0 <- log(100)
    ##
    eta.emax ~ .5
    eta.ec50  ~ .5
    eta.kout ~ .5
    eta.e0 ~ .5
    ##
    pdadd.err <- 10
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    alag1 <- talag
    ##
    #poplogit = log(temax/(1-temax))
    logit=exp(poplogit+eta.emax)
    #logit=temax+eta.emax
    emax = logit/(1+logit)
    ec50 =  exp(tec50 + eta.ec50)
    kout = exp(tkout + eta.kout)
    e0 = exp(te0 + eta.e0)
    ##
    DCP = center/v
    PD=1-emax*DCP/(ec50+DCP)
    ##
    effect(0) = e0
    kin = e0*kout
    ##
    d/dt(gut) =  -ka * gut
    d/dt(center) =  ka * gut - cl / v * center
    d/dt(effect) = kin*PD -kout*effect
    alag(gut) = alag1
    ##
    cp = center / v
    cp ~ prop(prop.err) + add(pkadd.err)
    pca = effect
    pca ~ add(pdadd.err) | pca
  })
}
nlmixr(pk.turnover.emax.lag)
#> i parameter labels from comments will be replaced by 'label()'
#> __ RxODE-based ODE model _______________________________________________________ 
#> -- Initialization: ------------------------------------------------------------- 
#> Fixed Effects ($theta): 
#>      talag        tka        tcl         tv   poplogit      tec50      tkout 
#>  0.1000000  0.0000000 -2.3025851  2.3025851  2.0000000 -0.6931472 -2.9957323 
#>        te0 
#>  4.6051702 
#> 
#> Omega ($omega): 
#>          eta.ka eta.cl eta.v eta.emax eta.ec50 eta.kout eta.e0
#> eta.ka        1      0     0      0.0      0.0      0.0    0.0
#> eta.cl        0      2     0      0.0      0.0      0.0    0.0
#> eta.v         0      0     1      0.0      0.0      0.0    0.0
#> eta.emax      0      0     0      0.5      0.0      0.0    0.0
#> eta.ec50      0      0     0      0.0      0.5      0.0    0.0
#> eta.kout      0      0     0      0.0      0.0      0.5    0.0
#> eta.e0        0      0     0      0.0      0.0      0.0    0.5
#> -- Multiple Endpoint Model ($multipleEndpoint): -------------------------------- 
#> +-----------------------------------------------------------------------------+
#> Â¦ variable                Â¦ cmt                     Â¦ dvid*                   Â¦
#> +-------------------------+-------------------------+-------------------------Â¦
#> Â¦ cp ~ ...                Â¦ cmt='cp' or cmt=4       Â¦ dvid='cp' or dvid=1     Â¦
#> +-------------------------+-------------------------+-------------------------Â¦
#> Â¦ pca ~ ...               Â¦ cmt='pca' or cmt=5      Â¦ dvid='pca' or dvid=2    Â¦
#> +-----------------------------------------------------------------------------+
#>   * If dvids are outside this range, all dvids are re-numered                  
#>   sequentially, ie 1,7, 10 becomes 1,2,3 etc                                   
#> -- mu-referencing ($muRefTable): ----------------------------------------------- 
#> +---------------------+
#> Â¦ tka      Â¦ eta.ka   Â¦
#> +----------+----------Â¦
#> Â¦ tcl      Â¦ eta.cl   Â¦
#> +----------+----------Â¦
#> Â¦ tv       Â¦ eta.v    Â¦
#> +----------+----------Â¦
#> Â¦ poplogit Â¦ eta.emax Â¦
#> +----------+----------Â¦
#> Â¦ tec50    Â¦ eta.ec50 Â¦
#> +----------+----------Â¦
#> Â¦ tkout    Â¦ eta.kout Â¦
#> +----------+----------Â¦
#> Â¦ te0      Â¦ eta.e0   Â¦
#> +---------------------+
#> -- Model: ---------------------------------------------------------------------- 
#>     ka <- exp(tka + eta.ka)
#>     cl <- exp(tcl + eta.cl)
#>     v <- exp(tv + eta.v)
#>     alag1 <- talag
#>     ##
#>     #poplogit = log(temax/(1-temax))
#>     logit=exp(poplogit+eta.emax)
#>     #logit=temax+eta.emax
#>     emax = logit/(1+logit)
#>     ec50 =  exp(tec50 + eta.ec50)
#>     kout = exp(tkout + eta.kout)
#>     e0 = exp(te0 + eta.e0)
#>     ##
#>     DCP = center/v
#>     PD=1-emax*DCP/(ec50+DCP)
#>     ##
#>     effect(0) = e0
#>     kin = e0*kout
#>     ##
#>     d/dt(gut) =  -ka * gut
#>     d/dt(center) =  ka * gut - cl / v * center
#>     d/dt(effect) = kin*PD -kout*effect
#>     alag(gut) = alag1
#>     ##
#>     cp = center / v
#>     cp ~ prop(prop.err) + add(pkadd.err)
#>     pca = effect
#>     pca ~ add(pdadd.err) | pca 
#> ________________________________________________________________________________

fit.TOF <- nlmixr(pk.turnover.emax.lag, warfarin, "focei", control=list(print=1),
                  table=list(cwres=TRUE, npde=TRUE));
#> i parameter labels from comments will be replaced by 'label()'
#> Needed Covariates:
#> [1] "CMT"
#> RxODE 0.9.2.2 using 4 threads (see ?getRxThreads)
#> Key: U: Unscaled Parameters; X: Back-transformed parameters; G: Gill difference gradient approximation
#> F: Forward difference gradient approximation
#> C: Central difference gradient approximation
#> M: Mixed forward and central difference gradient approximation
#> Unscaled parameters for Omegas=chol(solve(omega));
#> Diagonals are transformed, as specified by foceiControl(diagXform=)
#> |-----+---------------+-----------+-----------+-----------+-----------|
#> |    #| Objective Fun |     talag |       tka |       tcl |        tv |
#> |.....................|  prop.err | pkadd.err |  poplogit |     tec50 |
#> |.....................|     tkout |       te0 | pdadd.err |        o1 |
#> |.....................|        o2 |        o3 |        o4 |        o5 |
#> |.....................|        o6 |        o7 |...........|...........|
#> |-----+---------------+-----------+-----------+-----------+-----------|
#> Error in .fitFun(.ret) : 
#>   Evaluation error: Infinite/NaN while evaluating initial objective function.
#> Restart 1
#> Key: U: Unscaled Parameters; X: Back-transformed parameters; G: Gill difference gradient approximation
#> F: Forward difference gradient approximation
#> C: Central difference gradient approximation
#> M: Mixed forward and central difference gradient approximation
#> Unscaled parameters for Omegas=chol(solve(omega));
#> Diagonals are transformed, as specified by foceiControl(diagXform=)
#> |-----+---------------+-----------+-----------+-----------+-----------|
#> |    #| Objective Fun |     talag |       tka |       tcl |        tv |
#> |.....................|  prop.err | pkadd.err |  poplogit |     tec50 |
#> |.....................|     tkout |       te0 | pdadd.err |        o1 |
#> |.....................|        o2 |        o3 |        o4 |        o5 |
#> |.....................|        o6 |        o7 |...........|...........|
#> |-----+---------------+-----------+-----------+-----------+-----------|
#> Error in .fitFun(.ret) : 
#>   Evaluation error: Infinite/NaN while evaluating initial objective function.
#> Restart 2
#> Error in .fitFun(.ret) : 
#>   Not compatible with requested type: [type=list; target=double].
#> Restart 3
#> Error in .fitFun(.ret) : 
#>   Not compatible with requested type: [type=list; target=double].
#> Error in (function (data, inits, PKpars, model = NULL, pred = NULL, err = NULL, : Could not fit data.
```

<sup>Created on 2020-11-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This model runs and converges to reasonable estimates if I do any of the modifications I mentioned earlier. It will also run if talag is unbounded, initialized as `talag <- 0`, but it converges on -1.433e-14 and the other variables basically don't change.

The model also returns the same error even if talag is fixed to some value.

**Comment by mattfidler** [comment]
Thanks for reporting @jpryby.  I will check to see if this is still true on the development version.  The multiple endpoint handling in FOCEi has been rewritten (and should be more stable).  Also `alag`/`F` both support between subject variability.

**Comment by jpryby** [comment]
Sounds good. This is supposed to be the development version, or at least the foceiDur branch. I installed from your instructions here: https://github.com/nlmixrdevelopment/RxODE/issues/170#issuecomment-598264024

**Comment by mattfidler** [comment]
Even better.  I will look into it;  It looks like a `Rcpp` mismatch.

**Comment by jpryby** [comment]
Notably, the following workaround seems agreeable:

```
library(nlmixr)
pk.turnover.emax.lag <- function() {
    ini({
        talag <- c(0,0.1)
        tka <- log(1)
        tcl <- log(0.1)
        tv <- log(10)
        ##
        eta.ka ~ 1
        eta.cl ~ 2
        eta.v ~ 1
        prop.err <- 0.1
        pkadd.err <- 0.1
        ##
        poplogit <- 2
        #temax <- 7.5
        tec50 <- log(0.5)
        tkout <- log(0.05)
        te0 <- log(100)
        ##
        eta.emax ~ .5
        eta.ec50  ~ .5
        eta.kout ~ .5
        eta.e0 ~ .5
        ##
        pdadd.err <- 10
    })
    model({
        lateka <- exp(tka + eta.ka)
        cl <- exp(tcl + eta.cl)
        v <- exp(tv + eta.v)
        alag1 <- talag
        
        #alag workaround
        ka <- 0
        if (time >= alag1) {
            ka <- lateka
        }
        
        ##
        #poplogit = log(temax/(1-temax))
        logit=exp(poplogit+eta.emax)
        #logit=temax+eta.emax
        emax = logit/(1+logit)
        ec50 =  exp(tec50 + eta.ec50)
        kout = exp(tkout + eta.kout)
        e0 = exp(te0 + eta.e0)
        ##
        DCP = center/v
        PD=1-emax*DCP/(ec50+DCP)
        ##
        effect(0) = e0
        kin = e0*kout
        ##
        d/dt(gut) =  -ka * gut
        d/dt(center) =  ka * gut - cl / v * center
        d/dt(effect) = kin*PD -kout*effect
        #alag(gut) = alag1
        ##
        cp = center / v
        cp ~ prop(prop.err) + add(pkadd.err)
        pca = effect
        pca ~ add(pdadd.err) | pca
    })
}
```
For my specific model which fits multi-dose data I had to  derive a time-after-dose (TAD) variable, but this would be the same idea for the warfarin single-dose data. mtime would also work, but only for this single-dose.

**Comment by mattfidler** [comment]
Thanks @jpryby 

You could also use `RxODE`'s new `tad()` function.  

See the release notes:

https://github.com/nlmixrdevelopment/RxODE/blob/pruneBranch/NEWS.md

> Added `tad()`, `tad(cmt)` funtions for time since last dose and time since last dose for a compartment; Also added time after first dose and time after first dose for a compartment `tafd()`, `tafd(cmt)`; time of last dose `tlast()`, `tlast(cmt)` and dose number dosenum() (currently not for each compartment)

**Comment by mattfidler** [comment]
Hm.  The restart was having issues.  Perhaps more information about where the infinite objective function would be helpful.

**Comment by jpryby** [comment]
Trying another solution that uses mtime, but it returns this:
`Error: function 'mtime' or its derivatives are not supported in RxODE`

I updated to the latest pruneBranch RxODE as well.

**Comment by mattfidler** [comment]
How did you code `mtime()`?

**Comment by jpryby** [comment]
I tried a few different ways of coding it. One calculated dose time in the model like you do in one example:
```
if (!is.na(amt)){
        tdose <- time
    }
```
then defined `tabs <- tdose + ALAG1`, set `mtime(mt1) = tabs` and changed KA from 0 to the estimated value if `t >= tabs`. This updated mtime in a time-dependent fashion, so I assume that was part of the problem. However, this example also had to be changed to `amt>0` instead of `!is.na(amt)` since I got the same error `Error: function 'is.na' or its derivatives...`

**Comment by mattfidler** [comment]
`is.na` and family need to be changed in `RxODE`;  The derivatives of these functions need to be defined in order to be used.  this includes `tad()` etc.

**Comment by mattfidler** [comment]
This is too much information I believe

**Comment by mattfidler** [comment]
It appears the function `likInner0(eta, id)` is infinite for many subjects.

**Comment by mattfidler** [comment]
``` r
library(nlmixr)

pk.turnover.emax.lag <- function() {
  ini({
    talag <- c(0,0.1)
    tka <- log(1)
    tcl <- log(0.1)
    tv <- log(10)
    ##
    eta.ka ~ 1
    eta.cl ~ 2
    eta.v ~ 1
    prop.err <- 0.1
    pkadd.err <- 0.1
    ##
    poplogit <- 2
    #temax <- 7.5
    tec50 <- log(0.5)
    tkout <- log(0.05)
    te0 <- log(100)
    ##
    eta.emax ~ .5
    eta.ec50  ~ .5
    eta.kout ~ .5
    eta.e0 ~ .5
    ##
    pdadd.err <- 10
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    alag1 <- talag
    ##
    #poplogit = log(temax/(1-temax))
    logit=exp(poplogit+eta.emax)
    #logit=temax+eta.emax
    emax = logit/(1+logit)
    ec50 =  exp(tec50 + eta.ec50)
    kout = exp(tkout + eta.kout)
    e0 = exp(te0 + eta.e0)
    ##
    DCP = center/v
    PD=1-emax*DCP/(ec50+DCP)
    ##
    effect(0) = e0
    kin = e0*kout
    ##
    d/dt(gut) =  -ka * gut
    d/dt(center) =  ka * gut - cl / v * center
    d/dt(effect) = kin*PD -kout*effect
    alag(gut) = alag1
    ##
    cp = center / v
    cp ~ prop(prop.err) + add(pkadd.err)
    pca = effect
    pca ~ add(pdadd.err) | pca
  })
}

fit.TOF <- nlmixr(pk.turnover.emax.lag, warfarin, "focei", foceiControl(print=0),
                  table=list(cwres=TRUE, npde=TRUE))
#> ℹ parameter labels from comments will be replaced by 'label()'
#> Needed Covariates:
#> [1] "CMT"
#> RxODE 0.9.2.2 using 4 threads (see ?getRxThreads)
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:27 
#> done
#> Calculating residuals/tables
#> done.
#> Compiling NPDE model...done
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : initial ETAs were nudged; (can control by foceiControl(etaNudge=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod
#> = FALSE, : ETAs were reset to zero during optimization; (Can control by
#> foceiControl(resetEtaP=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : Using S matrix to calculate covariance, can check sandwich or R matrix
#> with $covRS and $covR
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : gradient problems with initial estimate and covariance; see $scaleInfo

print(fit.TOF)
#> ── nlmixr FOCEi (outer: nlminb) fit ────────────────────────────────────────────
#> 
#> 
#>           OBJF      AIC      BIC Log-likelihood Condition Number
#> FOCEi 1027.542 1951.236 2026.476      -957.6181         1787.219
#> ── Time (sec $time): ───────────────────────────────────────────────────────────
#> 
#> 
#>          setup optimize covariance table  npde    other
#> elapsed 0.7491 27.98724   27.98724 0.021 0.555 85.45842
#> ── Population Parameters ($parFixed or $parFixedDf): ───────────────────────────
#> 
#> 
#>             Est.     SE %RSE Back-transformed(95%CI) BSV(CV%) Shrink(SD)%
#> talag      0.497 0.0772 15.5    0.497 (0.346, 0.649)                     
#> tka       -0.881  0.681 77.3     0.414 (0.109, 1.57)     151.      48.3% 
#> tcl        -1.98  0.125  6.3    0.137 (0.108, 0.176)     27.0      2.73% 
#> tv          2.08  0.128 6.14       8.03 (6.25, 10.3)     21.0      9.60% 
#> prop.err   0.159                               0.159                     
#> pkadd.err  0.209                               0.209                     
#> poplogit    3.76   1.82 48.3     43 (1.22, 1.51e+03)     70.7      79.7% 
#> tec50     -0.101  0.284  282     0.904 (0.518, 1.58)     47.5      7.95% 
#> tkout      -2.81 0.0877 3.12 0.0603 (0.0508, 0.0716)     9.48      30.8% 
#> te0         4.72  0.082 1.74         112 (95.8, 132)     8.33      48.7% 
#> pdadd.err   3.58                                3.58                     
#>  
#>   Covariance Type ($covMethod): s
#>   Some strong fixed parameter correlations exist ($cor) :
#>          cor:tka,talag      cor:tcl,talag       cor:tv,talag cor:poplogit,talag 
#>             0.216             -0.189              0.528             -0.127  
#>    cor:tec50,talag    cor:tkout,talag      cor:te0,talag        cor:tcl,tka 
#>             0.257             0.0200             0.0554             -0.355  
#>         cor:tv,tka   cor:poplogit,tka      cor:tec50,tka      cor:tkout,tka 
#>             0.349             -0.579             0.0679              0.262  
#>        cor:te0,tka         cor:tv,tcl   cor:poplogit,tcl      cor:tec50,tcl 
#>            -0.244             -0.561            -0.0128             -0.260  
#>      cor:tkout,tcl        cor:te0,tcl    cor:poplogit,tv       cor:tec50,tv 
#>            0.0460              0.501            -0.0136              0.355  
#>       cor:tkout,tv         cor:te0,tv cor:tec50,poplogit cor:tkout,poplogit 
#>            0.0497             -0.270              0.438             -0.709  
#>   cor:te0,poplogit    cor:tkout,tec50      cor:te0,tec50      cor:te0,tkout 
#>            -0.330             -0.656             -0.529              0.437  
#>  
#> 
#>   No correlations in between subject variability (BSV) matrix
#>   Full BSV covariance ($omega) or correlation ($omegaR; diagonals=SDs) 
#>   Distribution stats (mean/skewness/kurtosis/p-value) available in $shrink 
#>   Minimization message ($message):  
#>     false convergence (8) 
#>   In an ODE system, false convergence may mean "useless" evaluations were performed.
#>   See https://tinyurl.com/yyrrwkce
#>   It could also mean the convergence is poor, check results before accepting fit
#>   You may also try a good derivative free optimization:
#>     nlmixr(...,control=list(outerOpt="bobyqa"))
#> ── Fit Data (object is a modified tibble): ─────────────────────────────────────
#> 
#> # A tibble: 483 x 43
#>   ID     TIME    DV  EVID CMT     PRED     RES     WRES  IPRED    IRES   IWRES
#>   <fct> <dbl> <dbl> <int> <fct>  <dbl>   <dbl>    <dbl>  <dbl>   <dbl>   <dbl>
#> 1 1       0.5   0       0 cp    0.0144 -0.0144 -0.00575 0.0105 -0.0105 -0.0501
#> 2 1       1     1.9     0 cp    2.33   -0.431  -0.171   1.73    0.173   0.501 
#> 3 1       2     3.3     0 cp    5.69   -2.39   -0.892   4.36   -1.06   -1.46  
#> # … with 480 more rows, and 32 more variables: CPRED <dbl>, CRES <dbl>,
#> #   CWRES <dbl>, eta.ka <dbl>, eta.cl <dbl>, eta.v <dbl>, eta.emax <dbl>,
#> #   eta.ec50 <dbl>, eta.kout <dbl>, eta.e0 <dbl>, ka <dbl>, cl <dbl>, v <dbl>,
#> #   alag1 <dbl>, logit <dbl>, emax <dbl>, ec50 <dbl>, kout <dbl>, e0 <dbl>,
#> #   DCP <dbl>, PD <dbl>, kin <dbl>, cp <dbl>, pca <dbl>, tad <dbl>,
#> #   dosenum <dbl>, gut <dbl>, center <dbl>, effect <dbl>, EPRED <dbl>,
#> #   ERES <dbl>, NPDE <dbl>
```

<sup>Created on 2020-11-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

You can try yourself, but I think this is fixed

**Comment by jpryby** [comment]
I installed the most recent RxODE@pruneBranch and nlmixr@foceiDur  and the error seems resolved. First time I tried it I got the same error, but maybe that was a cache issue (even though it said it was cleared). Seems to be working for my actual model too. Thanks!

**Comment by mattfidler** [comment]
Probably could use a pothoc test to prevent regression.

**Comment by jpryby** [comment]
There's another issue with the OFV from nlmixr being completely different for NONMEM, which I noticed with a posthoc test, now that you mention that. They're both FOCEI (obviously), and I haven't observed that discrepancy outside of this multiple endpoint model. Outside of the scope of this issue, but I think it bears mentioning.

**Comment by mattfidler** [comment]
Thanks @jpryby;  This is probably occurring because the assumptions of what happens when things become zero due to rounding errors.

Note, that you may not actually be comparing the objective functions between NONMEM  and nlmixr.  The only time that would occur with `posthoc` is if the NONMEM and nlmixr optimization functions gives the same `ETA` values.  You can compare nlmixr objective function to NONMEM's objective function without this assumption but it requires an extra step.

**Comment by mattfidler** [comment]
Once you get the `etas` make sure they are in the correct order and then you can give them to `foceiControl`

```R
foceiControl(maxOuterIterations=0, maxInnerIterations=0, etaMat=etaMat)
```

**Comment by mattfidler** [comment]
@jpryby `tad()` and related items are now supported in nlmixr/RxODE

However `mtime()` support is more complicated.  It is basically telling the system to advance but that requires adding `mtime()` at the very end of the model generation.  It takes more time;  Since I am trying to release soon, I took `mtime()` off the list.

For me, the current nlmixr runs a model that was crashing becuase of a memory leak.  This means the model you have internally may run with the development branch.

**Comment by jpryby** [comment]
Sounds good, Matt, thanks! I'll give it another try.

**Comment by mattfidler** [comment]
Thanks @jpryby;  Also fixed the flat speed in RxODE per core.  It should only affect `saem` in nlmixr, but not `focei`

**Comment by jpryby** [comment]
I'm getting compilation errors on the updates, unfortunately. Looks like the relevant error message is here:

`units/quantity.hpp:197:9: error: constexpr constructor does not have empty body`

I get this at the end: `make: *** [C:/PROGRA~1/R/R-36~1.1/etc/x64/Makeconf:215: ode_cmt1.o] Error 1`

Similar issue when updating RxODE

**Comment by mattfidler** [comment]
Can you post the value of Makevars.win

You can access it is R with:

```r
file.edit("~/.R/Makevars.win")
```

**Comment by mattfidler** [comment]
Also note the R version

**Comment by jpryby** [comment]
Absolutely:

```
# Makvars.win
CXX14=$(BINPREF)g++ $(M_ARCH)
CXX14STD=-std=c++1y
CXX14FLAGS=-O2 -Wall
```

```
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18362)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.6.1 tools_3.6.1   
```

**Comment by mattfidler** [comment]
Thanks;  I will have to setup the system to reproduce. I hope to do this today or tomorrow.

**Comment by jpryby** [comment]
Got that TMDD model to run, but it appears the remaining issue causing the crash was the size of the dataset, so you were right about memory leaks. After the compilation issues get resolved in the new version, that seems to be the only remaining problem I've run into test-wise (most other issues have workarounds, etc, as you know).

I was also able to reproduce the compilation issues in a linux docker container.

session info:
```
R version 3.6.2 (2019-12-12)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)
```

The end of the error output:
```
...
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_bounded.hpp:24:42: error: ‘stan::scalar_seq_view<int, void> high_vec’ has incomplete type
     scalar_seq_view<T_high> high_vec(high);
                                          ^
In file included from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/mat/err/check_pos_definite.hpp:6:0,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/mat/err/check_corr_matrix.hpp:7,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/mat.hpp:14,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/rev/mat.hpp:12,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math.hpp:4,
                 from lincmtB.cpp:3:
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_not_nan.hpp: In instantiation of ‘static void stan::math::internal::not_nan<T_y, is_vec>::check(const char*, const char*, const T_y&) [with T_y = stan::math::var; bool is_vec = false]’:
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_not_nan.hpp:49:78:   required from ‘void stan::math::check_not_nan(const char*, const char*, const T_y&) [with T_y = stan::math::var]’
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_2F1_converges.hpp:39:48:   required from ‘void stan::math::check_2F1_converges(const char*, const T_a1&, const T_a2&, const T_b1&, const T_z&) [with T_a1 = stan::math::var; T_a2 = stan::math::var; T_b1 = stan::math::var; T_z = stan::math::var]’
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/fun/grad_2F1.hpp:38:48:   required from ‘void stan::math::grad_2F1(T&, T&, const T&, const T&, const T&, const T&, const T&, int) [with T = stan::math::var]’
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/rev/scal/fun/grad_inc_beta.hpp:45:49:   required from here
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_not_nan.hpp:17:31: error: no matching function for call to ‘is_nan(double)’
     if (is_nan(value_of_rec(y))) {
                               ^
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/err/check_not_nan.hpp:17:31: note: candidate is:
In file included from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/fun/is_any_nan.hpp:4:0,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/rev/core/operator_addition.hpp:7,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/rev/core.hpp:18,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/rev/mat.hpp:6,
                 from /pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math.hpp:4,
                 from lincmtB.cpp:3:
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/fun/is_nan.hpp:19:13: note: template<class T, class> bool stan::math::is_nan(T)
 inline bool is_nan(T x) {
             ^
/pkpd/R-3.6.2/lib64/R/library/StanHeaders/include/stan/math/prim/scal/fun/is_nan.hpp:19:13: note:   template argument deduction/substitution failed:
make: *** [lincmtB.o] Error 1
ERROR: compilation failed for package ‘RxODE’
* removing ‘/pkpd/R-3.6.2/lib64/R/library/RxODE’
* restoring previous ‘/pkpd/R-3.6.2/lib64/R/library/RxODE’
Error: Failed to install 'RxODE' from GitHub:
  (converted from warning) installation of package ‘/tmp/RtmpcETpBs/file9241ed4beee/RxODE_0.9.2-2.tar.gz’ had non-zero exit status
```

**Comment by mattfidler** [comment]
It seems that the error is occurring in `is_nan`.  This function is not found in the `linCmtB` C++ file.  I think there is some issue with setting up the Stan Headers.  

I would suggest checking: 

https://github.com/stan-dev/rstan/wiki/Installing-RStan-on-Linux#c-toolchain-configuration

They suggest using `-O3` which should work with RxODE, but I haven't tested it.

Also there is a specific note on CentOS 7.0 (which is the container you used):

https://github.com/stan-dev/rstan/wiki/Installing-RStan-on-Linux#special-note-centos-70

**Comment by mattfidler** [comment]
Also, I believe I have fixed all the memory leaks in RxODE, but I need to run another run of all the tests in valgrind level 2 which will take a day or so.

**Comment by mattfidler** [comment]
All the tests in valgind show that RxODE is clean from memory issues.

**Comment by mattfidler** [comment]
Also `-O3` works fine with RxODE (and of course speeds up nlmixr/RxODE)

**Comment by mattfidler** [comment]
> `units/quantity.hpp:197:9: error: constexpr constructor does not have empty body`

This also doesn't come from RxODE, but units.  I'm unsure what is happening here.

I ran the nlmixr unit tests and focei is working fine by these tests.  SAEM has some issues that need to be taken care of before release.

**Comment by jpryby** [comment]
Nice, glad to hear. I am continuing to have issues installing the updates on both my Windows and Linux platforms, unfortunately. Not sure if it's an R version issue. I re-installed rstan per the link you sent, but no luck. I understand the problem is based in these other packages, but it's only apparent when trying to install the updates. I am using `units_0.6-7`, which seems to be the latest version. I also have the latest CRAN release of rstan. What versions of rstan and units are you using? That may help ensure my setup is consistent.

**Comment by mattfidler** [comment]
Here is the session info:
```r
> devtools::session_info()
─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.0.3 (2020-10-10)
 os       Pop!_OS 20.10               
 system   x86_64, linux-gnu           
 ui       X11                         
 language en_US:en                    
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/Chicago             
 date     2020-11-18                  

─ Packages ───────────────────────────────────────────────────────────────────
 ! package       * version    date       lib source        
   assertthat      0.2.1      2019-03-21 [1] CRAN (R 4.0.3)
   backports       1.2.0      2020-11-02 [1] CRAN (R 4.0.3)
   brew            1.0-6      2011-04-13 [1] CRAN (R 4.0.3)
   callr           3.5.1      2020-10-13 [1] CRAN (R 4.0.3)
   checkmate       2.0.0      2020-02-06 [1] CRAN (R 4.0.3)
   cli             2.1.0      2020-10-12 [1] CRAN (R 4.0.3)
   colorspace      2.0-0      2020-11-11 [1] CRAN (R 4.0.3)
   crayon          1.3.4      2017-09-16 [1] CRAN (R 4.0.3)
   data.table      1.13.2     2020-10-19 [1] CRAN (R 4.0.3)
   desc            1.2.0      2018-05-01 [1] CRAN (R 4.0.3)
   devtools        2.3.2      2020-09-18 [1] CRAN (R 4.0.3)
   diffobj         0.3.2      2020-10-05 [1] CRAN (R 4.0.3)
   digest          0.6.27     2020-10-24 [1] CRAN (R 4.0.3)
   dparser         0.1.8      2017-11-13 [1] CRAN (R 4.0.3)
   dplyr         * 1.0.2      2020-08-18 [1] CRAN (R 4.0.3)
   ellipsis        0.3.1      2020-05-15 [1] CRAN (R 4.0.3)
   fansi           0.4.1      2020-01-08 [1] CRAN (R 4.0.3)
   fastGHQuad      1.0        2018-09-30 [1] CRAN (R 4.0.3)
   fs              1.5.0      2020-07-31 [1] CRAN (R 4.0.3)
   generics        0.1.0      2020-10-31 [1] CRAN (R 4.0.3)
   ggplot2         3.3.2      2020-06-19 [1] CRAN (R 4.0.3)
   glue            1.4.2      2020-08-27 [1] CRAN (R 4.0.3)
   gtable          0.3.0      2019-03-25 [1] CRAN (R 4.0.3)
   lattice         0.20-41    2020-04-02 [4] CRAN (R 4.0.0)
   lazyeval        0.2.2      2019-03-15 [1] CRAN (R 4.0.3)
   lbfgsb3c        2020-3.2   2020-03-03 [1] CRAN (R 4.0.3)
   lifecycle       0.2.0      2020-03-06 [1] CRAN (R 4.0.3)
   lotri           0.2.2      2020-05-29 [1] CRAN (R 4.0.3)
   magrittr        1.5        2014-11-22 [1] CRAN (R 4.0.3)
   Matrix          1.2-18     2019-11-27 [4] CRAN (R 4.0.0)
   memoise         1.1.0      2017-04-21 [1] CRAN (R 4.0.3)
   minqa           1.2.4      2014-10-09 [1] CRAN (R 4.0.3)
   munsell         0.5.0      2018-06-12 [1] CRAN (R 4.0.3)
   n1qn1           6.0.1-10   2020-11-17 [1] CRAN (R 4.0.3)
   nlme            3.1-151    2020-10-31 [1] local         
 P nlmixr        * 1.1.1.9    2020-11-17 [?] local         
   pillar          1.4.6      2020-07-10 [1] CRAN (R 4.0.3)
   pkgbuild        1.1.0      2020-07-13 [1] CRAN (R 4.0.3)
   pkgconfig       2.0.3      2019-09-22 [1] CRAN (R 4.0.3)
   pkgload         1.1.0      2020-05-29 [1] CRAN (R 4.0.3)
   PreciseSums     0.4        2020-07-10 [1] CRAN (R 4.0.3)
   prettyunits     1.1.1      2020-01-24 [1] CRAN (R 4.0.3)
   processx        3.4.4      2020-09-03 [1] CRAN (R 4.0.3)
   ps              1.4.0      2020-10-07 [1] CRAN (R 4.0.3)
   purrr           0.3.4      2020-04-17 [1] CRAN (R 4.0.3)
   R6              2.5.0      2020-10-28 [1] CRAN (R 4.0.3)
   Rcpp            1.0.5      2020-07-06 [1] CRAN (R 4.0.3)
   RcppArmadillo   0.10.1.2.0 2020-11-16 [1] CRAN (R 4.0.3)
   rematch2        2.1.2      2020-05-01 [1] CRAN (R 4.0.3)
   remotes         2.2.0      2020-07-21 [1] CRAN (R 4.0.3)
   rex             1.2.0      2020-04-21 [1] CRAN (R 4.0.3)
   rlang           0.4.8      2020-10-08 [1] CRAN (R 4.0.3)
   rprojroot       2.0.2      2020-11-15 [1] CRAN (R 4.0.3)
   rstudioapi      0.13       2020-11-12 [1] CRAN (R 4.0.3)
   RxODE         * 1.0.0-0    2020-11-17 [1] local         
   scales          1.1.1      2020-05-11 [1] CRAN (R 4.0.3)
   sessioninfo     1.1.1      2018-11-05 [1] CRAN (R 4.0.3)
   symengine       0.1.5      2020-07-06 [1] CRAN (R 4.0.3)
   sys             3.4        2020-07-23 [1] CRAN (R 4.0.3)
   testthat      * 3.0.0      2020-10-31 [1] CRAN (R 4.0.3)
   tibble          3.0.4      2020-10-12 [1] CRAN (R 4.0.3)
   tidyselect      1.1.0      2020-05-11 [1] CRAN (R 4.0.3)
   units           0.6-7      2020-06-13 [1] CRAN (R 4.0.3)
   usethis         1.6.3      2020-09-17 [1] CRAN (R 4.0.3)
   vctrs           0.3.5      2020-11-17 [1] CRAN (R 4.0.3)
   waldo           0.2.3      2020-11-09 [1] CRAN (R 4.0.3)
   withr           2.3.0      2020-09-22 [1] CRAN (R 4.0.3)

[1] /home/matt/R/x86_64-pc-linux-gnu-library/4.0
[2] /usr/local/lib/R/site-library
[3] /usr/lib/R/site-library
[4] /usr/lib/R/library

 P ── Loaded and on-disk path mismatch.
```

**Comment by jpryby** [comment]
We're using the same units version, but I'm not seeing rstan. I think the current CRAN version of that has some issues, based on some googling. Thanks for your help!

**Comment by mattfidler** [comment]
No problem.

I am currently trying built the correct setup using R 3.6.1 in windows.  To make sure it isn't a R version issue, I am compiling all the sources (so it will take some time).

**Comment by mattfidler** [comment]
It seems that the toolchain with R 4.0 has fixed some gcc bugs that didn't work in C++14;  Therefore the stan toolchain doesn't work with the broken API.

**Comment by jpryby** [comment]
Yep, I agree. The bug is based in boost, so I downgraded to a pre-R 4.0 version (1.66.0-1) and got the latest RxODE pruneBranch installed.

**Comment by mattfidler** [comment]
Awesome.  I was trying to install with MRAN to make sure things are correct.  I also noticed there were issues with RcppArmadillo, but it may still have to do with boost.

**Comment by jpryby** [comment]
Can now confirm downgrading BH also allowed me to update n1qn1 and nlmixr@foceiDur.

**Comment by mattfidler** [comment]
I am going through testing and noticing that some of the example models are not working as well as the sympy nlmixr

I am looking through the details now

**Comment by jpryby** [comment]
Good to know. I didn't notice any of my models performing worse with symengine vs sympy, but that's probably because the sympy version couldn't parse our more complicated models (so no comparison possible).

Also, still having memory leak issues with big datasets, even after updating. I can reproduce below by generating an absurdly large dataset, but I get the error in normal datasets (~400) if there are many dosing and obs events.

``` r
library(RxODE)
#> RxODE 1.0.0.0 using 4 threads (see ?getRxThreads)

set.seed(32)

mod <- RxODE({
  C2 = centr/V2*(1+prop.err);
  C3 = peri/V3;
  CL =  TCl*exp(eta.Cl) ## This is coded as a variable in the model
  d/dt(depot) =-KA*depot;
  d/dt(centr) = KA*depot - CL*C2 - Q*C2 + Q*C3;
  d/dt(peri)  =                    Q*C2 - Q*C3;
})
#> ABC
#>

theta <- c(KA=2.94E-01, TCl=1.86E+01, V2=4.02E+01,  # central 
           Q=1.05E+01, V3=2.97E+02)                # peripheral

## the column names of the omega matrix need to match the parameters specified by RxODE
omega <- lotri(eta.Cl ~ 0.4^2)
omega
#>        eta.Cl
#> eta.Cl   0.16
sigma <- lotri(prop.err ~ 1)
sigma
#>          prop.err
#> prop.err        1

ev <- et(amount.units="mg", time.units="hours") %>%
  et(amt=10000, cmt="depot")

ev <- ev %>% et(0,48, length.out=100)

nsubjects <- 2000
sim  <- rxSolve(mod,theta,ev,omega=omega,sigma=sigma,nSub=nsubjects,returnType="data.table")

library(nlmixr)
library(data.table)
#> Warning: package 'data.table' was built under R version 3.6.3

mod.nm <- function() {
  ini({
    KA=c(0,2.94E-01)
    TCl=c(0,1.86E+01)
    V2=c(4.02E+01)
    Q=c(0,1.05E+01)
    V3=c(0,2.97E+02)
    prop.err=c(0,1)
    
    eta.Cl ~ 0.16
  })
  model({
    C2 = centr/V2;
    C3 = peri/V3;
    CL =  TCl*exp(eta.Cl) ## This is coded as a variable in the model
    d/dt(depot) =-KA*depot;
    d/dt(centr) = KA*depot - CL*C2 - Q*C2 + Q*C3;
    d/dt(peri)  =                    Q*C2 - Q*C3;
    conc = centr/V2
    conc ~ prop(prop.err)
  })
}

sim.dose <- data.table(
  id = 1:nsubjects,
  amt = 10000,
  time = 0,
  dv = 0,
  evid = 1
)
sim.obs <- sim[,`:=`(
  id = sim.id,
  amt = 0,
  time = as.numeric(time),
  dv = C2,
  evid=0
)][,c("id","amt","time","dv","evid")]
sim.data <- rbindlist(list(
  sim.dose,
  sim.obs
))[order(id,time)]

fit <- nlmixr(mod.nm,sim.data,"focei",control=list(print=5))
#> Error in .fitFun(.ret) : 
#>   'Calloc' could not allocate memory (2149978348 of 8 bytes)
#> Restart 1
#> Error in .fitFun(.ret) : 
#>   'Calloc' could not allocate memory (2149978348 of 8 bytes)
#> Restart 2
#> Error in .fitFun(.ret) : 
#>   'Calloc' could not allocate memory (2149978348 of 8 bytes)
#> Restart 3
#> Error in .fitFun(.ret) : 
#>   'Calloc' could not allocate memory (2149978348 of 8 bytes)
#> Error in (function (data, inits, PKpars, model = NULL, pred = NULL, err = NULL, : Could not fit data.
```

<sup>Created on 2020-11-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattfidler** [comment]
This isn't a memory leak *per se*, but rather insufficient memory to run the problem.  A memory leak occurs when the internals are not releasing memory causing memory usage to get larger and larger until the computer crashes or R crashes.

The example you have means you cannot load the entire problem in memory.  This will happen with RxODE/nlmixr since the model is currently run in memory instead of reading data and running.  Typically everything in R is loaded into memory so this is not unique to the `nlmixr`.  There are ways around this, but nlmixr currently doesn't support or know about them.

The memory limitation is more extreme in nlmixr since you load the data into R and then duplicate it into memory outside of R (to allow threaded execution). 

I was thinking of allowing an in-place ODE solving which should allow larger datasets to be run and parsed (at the cost of parallel processing or more complexity in parallel processing to make sure R doesn't destroy the memory), but that isn't on the release list right now.

**Comment by jpryby** [comment]
Got it, makes sense. I'll look into workarounds for the time being.

**Comment by mattfidler** [comment]
As a work-around, you might be able to do the following:
1. Delete the unneeded columns in your input dataset
2. Preprocess your `nlmixr` function by `f <- nlmixr(f)` and hence overwrite the original function
3. Remove all the other items in your R session
4. Run the model (see if it works)

**Comment by mattfidler** [comment]
You may even wish to enforce garbage collection, though I think R will do this when you try to allocate that much memory:
```r
gc()
```

**Comment by mattfidler** [comment]
It is only one focei model so far and it could be something as simple as a different order of operations causing rounding errors that add up. This happens in different platforms of NONMEM. It may be expected with a different Cas platform.
